testcases/kernel/syscalls/fcntl/fcntl27.c

```diff
@@ -1,192 +1,51 @@
-     * perform global setup for test
-     ***************************************************************/
-	setup();
-
-	expected_result = -1;
-
-    /***************************************************************
-     * check looping state if -c option given
-     ***************************************************************/
-	for (lc = 0; TEST_LOOPING(lc); lc++) {
+ * Basic test for fcntl(2) using F_SETLEASE & F_RDLCK argument.
+ */
 
-		tst_count = 0;
+#include "lapi/fcntl.h"
+#include "tst_test.h"
 
 #ifdef F_SETLEASE
-		/*
-		 * Call fcntl(2) with F_SETLEASE & F_RDLCK argument on fname
-		 */
-		TEST(fcntl(fd, F_SETLEASE, F_RDLCK));
-
-		/* check return code */
-		if (TEST_RETURN == expected_result) {
-			tst_resm(TPASS,
-				 "fcntl(fd, F_SETLEASE, F_RDLCK) succeeded");
-		} else {
-			tst_resm(TFAIL, "fcntl(%s, F_SETLEASE, F_RDLCK)"
-				 " failed with errno %d : %s", fname,
-				 TEST_ERRNO, strerror(TEST_ERRNO));
-		}
-#else
-		tst_resm(TINFO, "F_SETLEASE not defined, skipping test");
-#endif
-	}
 
-    /***************************************************************
-     * cleanup and exit
-     ***************************************************************/
-	cleanup();
-	tst_exit();
-
-}
-
-/***************************************************************
- * setup() - performs all ONE TIME setup for this test.
- ***************************************************************/
-void setup(void)
+static struct test_case
 {
-
-	tst_sig(NOFORK, DEF_HANDLER, cleanup);
-
-	TEST_PAUSE;
-
-	tst_tmpdir();
-
-	sprintf(fname, "tfile_%d", getpid());
-	if ((fd = open(fname, O_RDWR | O_CREAT, 0777)) == -1) {
-		tst_brkm(TBROK, cleanup,
-			 "open(%s, O_RDWR|O_CREAT,0777) Failed, errno=%d : %s",
-			 fname, errno, strerror(errno));
-	}
-}
-
-/***************************************************************
- * cleanup() - performs all ONE TIME cleanup for this test at
- *				 completion or premature exit.
- ***************************************************************/
-void cleanup(void)
+	int oflags;
+	mode_t mode;
+	int cmd;
+	char *cmd_desc;
+	int arg;
+	char *arg_desc;
+	int exp_errno;
+} tcases[] = {
+	{O_RDWR | O_CREAT, 0777, F_SETLEASE, "F_SETLEASE", F_RDLCK, "F_RDLCK", EAGAIN},
+	{O_WRONLY | O_CREAT, 0222, F_SETLEASE, "F_SETLEASE", F_RDLCK, "F_RDLCK", EAGAIN},
+};
+
+static void verify_fcntl(unsigned int nr)
 {
+	struct test_case *tc = &tcases[nr];
+	int fd = SAFE_OPEN("testfile", tc->oflags, tc->mode);
 
-	/* close the file we've had open */
-	if (close(fd) == -1) {
-		tst_resm(TWARN, "close(%s) Failed, errno=%d : %s", fname, errno,
-			 strerror(errno));
-	}
+	TST_EXP_FAIL(fcntl(fd, tc->cmd, tc->arg), tc->exp_errno, "fcntl(fd, %s, %s)", tc->cmd_desc, tc->arg_desc);
+	SAFE_CLOSE(fd);
+}
 
-	tst_rmdir();
+static struct tst_test test = {
+	.test = verify_fcntl,
+	.tcnt = ARRAY_SIZE(tcases),
+	.needs_tmpdir = 1,
+};
 
-}
+#else
+TST_TEST_TCONF("F_SETLEASE not defined, skipping test");
```

Comment:
`F_SETLEASE` is from Linux 2.4. It was defined in glibc long time ago, it was added for all archs in 2012 (glibc 2.17) 
in commit [`335e6931e2 ("Reorganization of bits/fcntl.h for Linux")`](https://sourceware.org/git/?p=glibc.git;a=commitdiff;h=335e6931e2b77627f4bc2eb007c5e15e8dbdeb98;hp=ac5c94d31d3259a32bfcc3be0ba198c72ff906db) and it was for some archs even before => we can safely remove this check.